crypto/random/rand_test.go

```diff
@@ -228,52 +226,47 @@ func TestShuffle(t *testing.T) {
 	require.NoError(t, err)
 
 	t.Run("basic uniformity", func(t *testing.T) {
-		listSize := 100
-		sampleSize := 80000
-		// the distribution of a particular element of the list, testElement
-		distribution := make([]float64, listSize)
-		testElement := rand.Intn(listSize)
-		// Slice to shuffle
-		list := make([]int, 0, listSize)
-		for i := 0; i < listSize; i++ {
-			list = append(list, i)
-		}
-
-		shuffleAndCount := func(t *testing.T) {
-			err = rng.Shuffle(listSize, func(i, j int) {
-				list[i], list[j] = list[j], list[i]
-			})
-			require.NoError(t, err)
-			has := make(map[int]struct{})
-			for j, e := range list {
-				// check for repetition
-				_, ok := has[e]
-				require.False(t, ok, "duplicated item")
-				has[e] = struct{}{}
-				// fill the distribution
-				if e == testElement {
-					distribution[j] += 1.0
-				}
+		// compute n!
+		fact := func(n int) int {
+			f := 1
+			for i := 1; i <= n; i++ {
+				f *= i
 			}
+			return f
 		}
 
-		t.Run("shuffle a random permutation", func(t *testing.T) {
-			for k := 0; k < sampleSize; k++ {
-				shuffleAndCount(t)
-			}
-			EvaluateDistributionUniformity(t, distribution)
-		})
-
-		t.Run("shuffle a same permutation", func(t *testing.T) {
-			for k := 0; k < sampleSize; k++ {
-				// reinit the permutation to the same value
+		for listSize := 2; listSize <= 6; listSize++ {
+			factN := uint64(fact(listSize))
+			t.Logf("permutation size is %d (factorial is %d)", listSize, factN)
+			t.Run("shuffle a random permutation of size", func(t *testing.T) {
+				list := make([]int, 0, listSize)
 				for i := 0; i < listSize; i++ {
-					list[i] = i
+					list = append(list, i)
 				}
-				shuffleAndCount(t)
-			}
-			EvaluateDistributionUniformity(t, distribution)
-		})
+				permEncoding := func() (uint64, error) {
+					err = rng.Shuffle(listSize, func(i, j int) {
+						list[i], list[j] = list[j], list[i]
+					})
+					return uint64(EncodePermutation(list)), err
+				}
+				BasicDistributionTest(t, factN, 1, permEncoding)
+			})
+
+			t.Run("shuffle a same permutation", func(t *testing.T) {
```

Comment:
I don't understand why we have this test, which looks identical to the previous one.

Reply:
- the first test shuffles an array multiple times. The resulting array from the previous shuffle is shuffled again. The distribution is computed based on all intermediate arrays from each loop.
- the second test shuffles the same array [0,1, .. , n-1] multiple times. The resulting array after each loop is used for the distribution, but it is then ignored. The next loop shuffle is computed using the same starting array. 

The shuffling should be uniform in both scenarios. Some buggy implementations may be uniform in one scenario and not the other.